src/hotspot/cpu/x86/interp_masm_x86.cpp

```diff
@@ -73,13 +76,14 @@ void InterpreterMacroAssembler::profile_obj_type(Register obj, const Address& md
   testptr(obj, TypeEntries::type_unknown);
   jccb(Assembler::notZero, next); // already unknown. Nothing to do anymore.
 
-  cmpptr(mdo_addr, 0);
-  jccb(Assembler::equal, none);
-  cmpptr(mdo_addr, TypeEntries::null_seen);
-  jccb(Assembler::equal, none);
-  // There is a chance that the checks above (re-reading profiling
-  // data from memory) fail if another thread has just set the
+  // is_type_none?
+  testptr(rscratch1, TypeEntries::type_mask);
+  jccb(Assembler::zero, none);
+
+  // There is a chance that the checks above
+  // fail if another thread has just set the
   // profiling to this obj's klass
+  xorptr(obj, rscratch1); // get back original value before XOR
   xorptr(obj, mdo_addr);
```

Comment:
Was it bug here originally? We had 2 xors (including this) and now 3 for `obj`.

Reply:
Yes, there was a bug here originally.  `obj` could have contained a bad value.  The code wants to store the Klass* from `load_klass` with the low `TypeEntries::null_seen` bit from `mdo_addr` ORed in.  However, because the low bits of the Klass* are 0, XOR was used to both OR the low bits and SET the high bits, assuming the high bits from `mdo_addr` are 0.
The problem comes from assuming the `mdo_addr` memory value is stable and reloading the value from memory to compare against 0 and TypeEntries::null_seen.  A transient 0 can appear because of non-atomic updates like this:
`orptr(mdo_addr, TypeEntries::null_seen);`
An alternative fix would be to get rid of the races and use atomic compare-and-swap to write new values.

The new 2nd XOR is an optimization, instead of calling load_klass() again, or saving the original `obj` value somewhere else.  The 3nd XOR is a repeat of the 1st XOR.  I don't think this slow path retry is that useful, and not all platforms do it, but I left it in because I wanted to minimize changes.